File: imhere-api-gateway/src/main/java/com/kdongsu5509/imhereapigateway/filter/logger/BodyMasker.java
package com.kdongsu5509.imhereapigateway.filter.logger;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Component;

@Component
public class BodyMasker {
    private final ObjectMapper objectMapper;

    public BodyMasker(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    public String mask(String body) {
        String formatted = formatJson(body);
        return maskSensitiveFields(formatted);
    }

    private String formatJson(String body) {
        try {
            var jsonNode = objectMapper.readTree(body);
            return objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(jsonNode);
        } catch (Exception e) {
            return body;
        }
    }

    private String maskSensitiveFields(String json) {
        return json.replaceAll(
                "\"(password|pw|confirmPassword|secret)\"\\s*:\\s*\"[^\"]+\"",
                "\"$1\": \"*****\""
        );
    }
}

File: src/main/kotlin/com/kdongsu5509/support/logger/BodyMasker.kt
package com.kdongsu5509.support.logger

import org.springframework.stereotype.Component
import tools.jackson.databind.json.JsonMapper

@Component
class BodyMasker(private val jsonMapper: JsonMapper) {

    companion object {
        private val SENSITIVE_PATTERN =
            Regex(""""(password|pw|confirmPassword|secret)"\s*:\s*"[^"]+"""")
    }

    fun mask(body: String): String {
        val formatted = formatJson(body)
        return maskSensitiveFields(formatted)
    }

    private fun formatJson(body: String): String {
        return try {
            val jsonNode = jsonMapper.readTree(body)
            jsonMapper.writerWithDefaultPrettyPrinter().writeValueAsString(jsonNode)
        } catch (_: Exception) {
            body
        }
    }

    private fun maskSensitiveFields(json: String): String =
        json.replace(SENSITIVE_PATTERN, "\"$1\": \"*****\"")
}
